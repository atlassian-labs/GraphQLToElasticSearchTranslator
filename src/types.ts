export type ElasticSearchConfig = {
    description: string;
    db_field_name: string;
    db_type: "keyword" | "text" | "boolean" | "long" | "nested" | "match_phrase" | "None";
    type: "string" | "array";

    query_type?: "phrase" | "cross_fields";
    prefix_type?: "reverse";
    nested_type?: "text" | "keyword" | "match_phrase";
    aggs_field?: string;
    aggs_name?: string;
    aggs_type?: string;

    // Used in MultiMatch
    fields?: string[];
    operator?: any;
};

export type FilterConfig = {
    filter: {
        isNull: string[];
        isNotNull: string[];
        arguments: {
            [key: string]: ElasticSearchConfig;
        };
    };

    aggregate?: {
        description: string;
        db_field_mappings: { [k: string]: string };
    };
};

export type WhereFilterObject = {
    and?: WhereFilterObject[];
    or?: WhereFilterObject[];
    not?: WhereFilterObject[];
    isNull?: string;
    isNotNull?: string;
    [k: string]: any;
};
