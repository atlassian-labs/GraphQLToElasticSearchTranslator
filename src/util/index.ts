import type { WhereFilterObject } from "../types";

function getDepth(obj: WhereFilterObject): number {
    let depth = 0;

    if (obj.and) {
        obj.and.forEach((d) => {
            const tmpDepth = getDepth(d);
            if (tmpDepth > depth) {
                depth = tmpDepth;
            }
        });
    } else if (obj.or) {
        obj.or.forEach((d) => {
            const tmpDepth = getDepth(d);
            if (tmpDepth > depth) {
                depth = tmpDepth;
            }
        });
    }

    return 1 + depth;
}

class Stack<T> {
    items: T[];

    constructor() {
        this.items = [];
    }

    pop() {
        if (this.items.length === 0) {
            return "underflow";
        }
        const res = this.items.pop();
        if (res !== undefined) {
            return res;
        }
        return "underflow";
    }

    push(item: T) {
        this.items.push(item);
    }

    isEmpty() {
        return this.items.length === 0;
    }
}

export { Stack, getDepth };
