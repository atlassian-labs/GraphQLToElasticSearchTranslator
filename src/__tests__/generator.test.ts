import { QueryGenerator } from "../index";
import { FilterConfig } from "../types";
import { Stack } from "../util";
const bodybuilder = require('bodybuilder');


const testFilterConfig: FilterConfig  = {
    filter: {
        isNull: ['id'],
        isNotNull: ['email'],
        arguments: {
            id: {
                description: 'Unique Identifier',
                db_field_name: 'id',
                db_type: 'keyword',
                type: 'string'
            },
            email: {
                description:
                    'This endpoint takes a contact (email) and returns all link documents for that contact.',
                db_field_name: 'email.raw',
                db_type: 'keyword',
                type: 'string'
            },
            and: {
                description: 'AND operator',
                db_field_name: 'None',
                db_type: 'None',
                type: 'array'
            },
            or: {
                description: 'OR operator',
                db_field_name: 'None',
                db_type: 'None',
                type: 'array'
            },
            not: {
                description: 'NOT operator',
                db_field_name: 'None',
                db_type: 'None',
                type: 'array'
            },
            isNotNull: {
                description: 'field name is not null',
                db_field_name: 'None',
                db_type: 'None',
                type: 'string'
            },
            isNull: {
                description: 'field name is null',
                db_field_name: 'None',
                db_type: 'None',
                type: 'string'
            }
        }
    }
};

describe('Elasticsearch QueryGenerator tests', () => {
    it("Default QueryGenerator Constructor", () => {
        const qg = new QueryGenerator();
        expect(qg.filterDepth).toBe(0);
        expect(qg.maxFilterDepth).toBe(3);
    
        expect(qg.nestedPath).toBeInstanceOf(Stack);
        expect(qg.nestedPath.isEmpty()).toBe(true);
        expect(qg.nestedDict).toMatchObject({});
    });

    it('should return correctly formatted queries given a email', () => {
        let args = {
            email:'test@test.com'
        };
        let query = bodybuilder();
        let boolquery = new QueryGenerator().getFilterQuery(
            query,
            args,
            testFilterConfig
        );
        expect(boolquery.build()).toMatchSnapshot();
    });

    it('should return correctly formatted queries given an AND operator', () => {
        let args = {
            and: [{ id: '123' }, { email: 'test@test.com' }]
        };
        let query = bodybuilder();
        let boolquery = new QueryGenerator().getFilterQuery(
            query,
            args,
            testFilterConfig
        );
        expect(boolquery.build()).toMatchSnapshot();
    });

    it('should return correctly formatted queries given an OR operator', () => {
        let args = {
            or: [{ id: '123' }, { email: 'test@test.com' }]
        };
        let query = bodybuilder();
        let boolquery = new QueryGenerator().getFilterQuery(
            query,
            args,
            testFilterConfig
        );
        expect(boolquery.build()).toMatchSnapshot();
    });

    it('should return correctly formatted queries given an isNull operator', () => {
        let args = {
            isNull: 'id'
        };
        let query = bodybuilder();
        let boolquery = new QueryGenerator().getFilterQuery(
            query,
            args,
            testFilterConfig
        );
        expect(boolquery.build()).toMatchSnapshot();
    });

    it('should return correctly formatted queries given an isNull operator', () => {
        let args = {
            isNotNull: 'email'
        };
        let query = bodybuilder();
        let boolquery = new QueryGenerator().getFilterQuery(
            query,
            args,
            testFilterConfig
        );
        expect(boolquery.build()).toMatchSnapshot();
    });
});