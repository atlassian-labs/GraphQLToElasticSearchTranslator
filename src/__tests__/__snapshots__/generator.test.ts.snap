// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Elasticsearch QueryGenerator tests should return correctly formatted queries given a email 1`] = `
Object {
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "term": Object {
            "email.raw": "test@test.com",
          },
        },
      ],
    },
  },
}
`;

exports[`Elasticsearch QueryGenerator tests should return correctly formatted queries given an AND operator 1`] = `
Object {
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "term": Object {
            "id": "123",
          },
        },
        Object {
          "term": Object {
            "email.raw": "test@test.com",
          },
        },
      ],
    },
  },
}
`;

exports[`Elasticsearch QueryGenerator tests should return correctly formatted queries given an OR operator 1`] = `
Object {
  "query": Object {
    "bool": Object {
      "should": Array [
        Object {
          "term": Object {
            "id": "123",
          },
        },
        Object {
          "term": Object {
            "email.raw": "test@test.com",
          },
        },
      ],
    },
  },
}
`;

exports[`Elasticsearch QueryGenerator tests should return correctly formatted queries given an isNull operator 1`] = `
Object {
  "query": Object {
    "bool": Object {
      "must_not": Array [
        Object {
          "exists": Object {
            "field": "id",
          },
        },
      ],
    },
  },
}
`;

exports[`Elasticsearch QueryGenerator tests should return correctly formatted queries given an isNull operator 2`] = `
Object {
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "exists": Object {
            "field": "email.raw",
          },
        },
      ],
    },
  },
}
`;
